### DISCUSSION TITLE
[Pinned] Installation & FAQ — read this first

### DISCUSSION THREAD

**Discussion by nankingjing** [nankingjing/PoseMamba#20] [Announcements]
# PoseMamba — Installation & FAQ

> **Official troubleshooting thread** · AAAI 2025 · [Paper](https://ojs.aaai.org/index.php/AAAI/article/view/32401) · [README](https://github.com/nankingjing/PoseMamba#readme)

Before opening a new issue, please check this thread. Most past issues are covered here.

---

## 1. Environment setup (recommended)

```bash
conda create -n posemamba python=3.8.5
conda activate posemamba
pip install torch==1.13.1+cu117 torchvision==0.14.1+cu117 torchaudio==0.13.1 --extra-index-url https://download.pytorch.org/whl/cu117
pip install -r requirements.txt
cd kernels/selective_scan && pip install -e . && cd ../..
```

**You do NOT need to clone the official Mamba repo.** Use `kernels/selective_scan` included in this repository.

---

### Q1: `selective_scan_cuda_core is not defined`

CUDA extension not built correctly. Rebuild:

```bash
cd kernels/selective_scan
pip uninstall selective-scan -y 2>/dev/null || true
pip install -e .
cd ../..
```

Ensure PyTorch/CUDA versions match (tested: torch 1.13.1 + cu117). On CUDA 12.x, try editing `setup.py`:

```python
MODES = ["core", "oflex"]
```

---

### Q2: `kernels/selective_scan` not found

The folder is in the repo root. Re-clone the full repository — do not use incomplete zip downloads.

```bash
ls kernels/selective_scan/setup.py
```

---

### Q3: Dataset not found (`h36m_sh_conf_cam_source_final.pkl`)

1. Download preprocessed H36M data from [README links](https://github.com/nankingjing/PoseMamba#human36m) → `data/motion3d/`
2. Run slicing:

```bash
cd tools && python convert_h36m.py && cd ..
```

3. Verify:

```bash
ls data/motion3d/MB3D_f243s81/h36m_sh_conf_cam_source_final.pkl
```

---

### Q4: Evaluation command / checkpoint format

**There is no `--eval-only` flag.** Use `--evaluate`:

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --config checkpoint/pose3d/PoseMamba_B/config.yaml \
  --evaluate checkpoint/pose3d/PoseMamba_B/best_epoch.bin \
  --checkpoint eval/checkpoint
```

| Source | Format | Notes |
|--------|--------|-------|
| Training output | `best_epoch.bin` | dict with `model_pos` key |
| Google Drive release | `.pth.tr` | load with `torch.load()` — **do not unzip** |

If `config.yaml` is missing in checkpoint dir (pretrained-only eval):

```bash
mkdir -p checkpoint/pose3d/PoseMamba_L
cp configs/pose3d/PoseMamba_train_h36m_L.yaml checkpoint/pose3d/PoseMamba_L/config.yaml
```

See `eval.sh` for S/B/L examples.

---

### Q5: `_pickle.UnpicklingError` when loading weights

Usually **PyTorch version mismatch**. Use README-tested torch 1.13.1, or:

```python
torch.load('weights.pth', map_location='cpu')
```

---

### Q6: Multi-GPU / distributed training

**Not officially supported.** Use single GPU:

```bash
CUDA_VISIBLE_DEVICES=0 python train.py --config configs/pose3d/PoseMamba_train_h36m_S.yaml --checkpoint checkpoint/pose3d/PoseMamba_train_h36m_S
```

`nn.DataParallel` with custom CUDA ops often causes device mismatch errors.

---

### Q7: Results gap vs paper (~1–2 mm)

Checklist:
- [ ] Correct model variant (S / B / L) and matching config
- [ ] Flip test enabled (`flip: True` in config)
- [ ] Same metric protocol (detected 2D vs GT 2D)
- [ ] `kernels/selective_scan` built on current machine
- [ ] Full 120 epochs if training from scratch

Reopen an issue with: config yaml, eval command, and MPJPE log if gap persists.

---

### Q8: Training time per epoch

With default `batch_size=4`, clip_len=243: **~10–15 min/epoch** on 3090/A100 is normal. Speed up with larger batch (if VRAM allows), `num_workers=4–8`, and SSD for data.

---

### Q9: MACs computation

We use **`thop`** (see `lib/model/PoseMamba.py`). Paper MACs = thop FLOPs ÷ 2, consistent with MotionBERT / MotionAGFormer reporting.

---

### Q10: COCO / custom 2D training

PoseMamba is **2D-to-3D lifting** (H36M / 3DHP). For wild video, `vis.py` uses an external 2D detector + PoseMamba 3D lifting. Native COCO 2D-only training is out of current scope.

---

## 3. Quick links

- **Demo**: `python vis.py --video sample_video.mp4 --gpu 0`
- **Configs**: `configs/pose3d/` (not `configs/h36m`)
- **Pretrained weights**: [Google Drive](https://drive.google.com/file/d/1WFRAeal8W6ntrTPNrf-SNywdgupj0-S8/view?usp=sharing)
- **Author**: [Google Scholar](https://scholar.google.com/citations?user=u2QDgXkAAAAJ&hl=zh-CN)

---

## 4. Still stuck?

Open a [new issue](https://github.com/nankingjing/PoseMamba/issues/new) with:

1. OS + GPU
2. `python -V` / `torch.__version__` / `nvcc --version`
3. Full command and traceback
4. Config yaml (or checkpoint path)

We'll respond as soon as possible. Thanks for using PoseMamba!
